src/main/java/org/springframework/integration/kafka/listener/QueueingMessageListenerInvoker.java

```diff
@@ -94,63 +105,69 @@ public void enqueue(KafkaMessage message) {
 	}
 
 	public void start() {
+
+		if(ringBufferProcessor != null){
+			stop(-1l);
+		}
+		ExecutorService service = executorService != null ? executorService : Executors.newSingleThreadExecutor();
 		this.running = true;
+		this.ringBufferProcessor = RingBufferProcessor.share(service, capacity);
+		this.ringBufferProcessor.subscribe(new KafkaMessageDispatchingSubscriber());
 	}
 
 	public void stop(long stopTimeout) {
-		shutdownLatch = new CountDownLatch(1);
 		this.running = false;
-		try {
-			shutdownLatch.await(stopTimeout, TimeUnit.MILLISECONDS);
-		}
-		catch (InterruptedException e) {
-			Thread.currentThread().interrupt();
+		if(ringBufferProcessor != null) {
+			ringBufferProcessor.onComplete();
```

Comment:
Shutdown latch must be set before `onComplete` or else we risk blocking needlessly below.